### PR TITLE
DHCPv6 Relay Agent

### PR DESCRIPTION
**Pull request by kellyyeh**

### CODE REVIEW

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+options: include a &quot;Relay Message option&quot;
+
+# DHCPv6 Packet Forwarding
+
+The DHCPv6 relay agent on the routing switch forwards DHCPv6 client packets to all DHCPv6 servers that are configured in the table administrated for each VLAN.
+
+A DHCPv6 client locates a DHCPv6 server using a reserved, link-scoped multicast address.
+
+The packets are forwarded using multicasting to a group of DHCPv6 servers
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:99]: packet are forwarded to configurable IPv6 helpers addresses.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+Currently, OneBoot team uses server blade MAC address as unique key to keep track of IP resources. With heterogenous DHCP client implementation across the network, DUIDs could not resolve IP resource tracking issue. The two types of DUIDs, DUID-LL and DUID-LLT used to facilitate resource tracking both have link layer addresses embedded. The current client link-layer address option in DHCPv6 specification limits the DHCPv6 Relay to first hop to provide the client link layer address, which are relay agents that are connected to the same link as the client, and that limits SONiC DHCPv6 deployment to ToR/MoR switches for early stages. One olution would be to provide SONiC&#39;s own DHCPv6 relay agent feature.
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:44]: Please update section as per internal discussion.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+9. DHCPv6 client receives the REPLY message that contains the desired IP address.
+
+![image](https://user-images.githubusercontent.com/42761586/117859842-5fd13b00-b244-11eb-9297-c2674d128dd9.png)
+
+# Requirements
+
+- Configured and running DHCPv6 client and server
+- Connectivity between the relay agent and DHCPv6 server
+- Configure one or more IP helper addresses for specified VLANs to forward DHCPv6 requests to DHCPv6 servers on other subnets.
+- IPv6 multicast address for all DHCP agents: IPv6 ff02::1:2
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:124]: nit: ALL DHCPv6 Servers and agents multicast address.

Comment (tahmed-dev) [reply doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:124]: Why is this a requirement? please provide details.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+# Topology
+
+![image](https://user-images.githubusercontent.com/42761586/117859824-58aa2d00-b244-11eb-991f-7ad8759f6612.png)
+
+# Design
+
+# CLI and Usage
+
+-show dhcprelay counters
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:136]: should it be `dhcp6relay counters`?

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+-clear dhcprelay counters
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:138]: we may extend `sonic-clear` with `-dhcp6 counter`. @lguohan, @jleveque do you have preference here?

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+# Source IP
+
+VLAN SVI IP
+
+Configurable option to use loopback address for dual ToR
+
+# Multiple VLAN
+
+Check kernel routing table and find mapping between interface and server
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:199]: the `link_address` field should help select the appropriate vlan interface. No need to check the kernel routing here.

**Review by tahmed-dev — COMMENTED** [review]
Thanks for your work,

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md (hunk 5 above)
Comment (jleveque) [reply doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:138]: Yes, we should extend `sonic-clear`. @kellyyeh: Please look at the existing subcommand structure of `sonic-clear` and align with that.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md (hunk 3 above)
Comment (kellyyeh) [reply doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md]: This is not a requirement. I mistakenly thought multicasting was used for our relay agent. Made the revision!

**Review by tahmed-dev — DISMISSED** [review]
LGTM, few minor comments. Thanks for your work.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+RELEASE
+DECLINE
+RELAY-FORWARD
+RELAY-REPLY
+
+# CONFIG DB schema
+
+DHCP|intf-i|dhcpv6\_servers: [&quot;dhcp-server-0&quot;, &quot;dhcp-server-1&quot;, ...., &quot;dhcp-server-n-1&quot;]
+
+DHCP|intf-i|dhcpv6\_options: []
```
Comment (lguohan) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:165]: need yang model.

**Review by tahmed-dev — APPROVED** [review]
LGTM, thanks!

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+# Overview
+
+SONiC currently supports DHCPv4 Relay via the use of open source ISC DHCP package. However, DHCPv6 specification does not define a way to communicate client link-layer address to the DHCP server where DHCP server is not connected to the same network link as DHCP client. DHCPv6 requires all clients prepare and send a DUID as the client identifier in all DHCPv6 message exchanges. However, these methods do not provide a simple way to extract a client's link-layer address. Providing option 79 in DHCPv6 Relay-Forward messages will help carry the client link-layer address explicitly. The server needs to know the client's MAC address to allow DHCP Reservation, which provides pre-set IP address to specific client based on its physical MAC address. The DHCPv6 relay agent is able to read the source MAC address of DHCPv6 messages that it received from client, and encapsulate these messages within a DHCPv6 Relay-Forward message, inserting the client MAC address as option 79 in the Relay-Forward header sent to the server.
```
Comment (venkatmahalingam) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:42]: Please add details on why we think opensource ISC package is not a right fit for this DHCP6 relay enhancements.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+- DHCP server: receives clients' requests and replies to them
+- DHCP client: send configuration requests to the server
+- DHCP relay agent: forwards DHCP packets between clients and servers that do not reside on a shared physical subnet
+
+1. Solicit: DHCPv6 client sends a SOLICIT message to locate DHCPv6 servers to the All\_DHCP\_Relay\_Agents\_and\_Servers multicast address.
+2. Advertise: DHCPv6 server sends an ADVERTISE message to indicate that it is available for DHCP service, in response to the SOLICIT message
+3. Request, Renew, Rebind: DHCPv6 client sends a REQUEST message to request configuration parameters(IP address or delegated prefixes) from the DHCPv6 server
+4. Reply: DHCPv6 server sends a REPLY message containing assigned addresses and configuration parameters in response to a CONFIRM message that confirms or denies that the addresses assigned to the client are appropriate to the link to which the client is connected. REPLY message acknowledges receipt of a RELEASE or DECLINE message.
+
+![image](https://user-images.githubusercontent.com/42761586/117859723-3adcc800-b244-11eb-9dd4-dbde609185a1.png)
```
Comment (lguohan) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:59]: please upload the image to this folder images directory.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+        list VLAN_LIST {
+    		key name;
+   		    leaf name {
+    			type string;
+  		    }
+   		    leaf dhcpv6_servers {
+     		    type inet6:ip-address;
+  		    }
+		    leaf options {
+			    type uint16;
```
Comment (lguohan) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:184]: does this only allow one option, what if we have other options? I suggest to define option79 true/false. some options needs parameter. it is difficult to define a universal way to define options.

Comment (kellyyeh) [reply doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md]: Updated option 79 to have a bool type! Future options added can define its own parameter.

File: doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md [hunk 11]
```diff
@@ -165,7 +165,7 @@ RELAY-REPLY
 <pre>
 DHCP|intf-i|dhcpv6_servers: [&quot;dhcp-server-0&quot;, &quot;dhcp-server-1&quot;, ...., &quot;dhcp-server-n-1&quot;]
 
-DHCP|intf-i|dhcpv6_option|link_layer_addr: [&quot;79&quot;]
+DHCP|intf-i|dhcpv6_option|link_layer_addr: &quot;true&quot;
```
Comment (tahmed-dev) [inline doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md:168]: what about naming this option rfc6939_support?

Comment (kellyyeh) [reply doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md]: Updated!

**Review by tahmed-dev — APPROVED** [review]

**Pull request merged**